### PR TITLE
[hlm-707]: Generic Row Mapper with test cases.

### PR DESCRIPTION
**Pull request by shivab-egov**
Generic Row Mapper.

### CODE REVIEW

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.HashMap;
+
+public class GenericRowMapper <T> implements RowMapper<T> {
+    private Class<T> mappedClass;
+
+    public GenericRowMapper(Class<T> mappedClass) {
+        this.mappedClass = mappedClass;
+    }
+
+    private Object instantiate(Class<?> clazz, HashMap row) throws IllegalAccessException, IllegalArgumentException {
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:22]: rename to mapRow

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        Object mappedObject = BeanUtils.instantiateClass(clazz);
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:23]: rename to parentObject

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        Field[] fields = mappedObject.getClass().getDeclaredFields();
+        for (Field f : fields) {
+            f.setAccessible(true);
+            if (!f.getType().isPrimitive() && !DataUtils.isWrapper(f)) {
+                Object mappedNestedObject = instantiate(f.getType(), row);
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:28]: rename to nestedObject

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                f.set(mappedObject, mappedNestedObject);
+            }else{
+                Object value = row.get(f.getName());
+                f.set(mappedObject, value);
+            }
+        }
+        return mappedObject;
+    }
+
+    private HashMap getMap(ResultSet rs) throws SQLException, ClassNotFoundException {
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:38]: Change to Map<String, Object>

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        ResultSetMetaData meta_data = rs.getMetaData();
+        int columnCount = meta_data.getColumnCount();
+
+        HashMap row = new HashMap(columnCount);
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:42]: Change to Map<String, Object> row = new HashMap(columnCount);

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            row.put(column, value);
+        }
+        return row;
+    }
+
+    @Override
+    public T mapRow(ResultSet rs, int rowNum) throws SQLException{
+        T instance;
+        try {
+            HashMap row = getMap(rs);
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:56]: Change to Map<String, Object> row

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            }else{
+                Object value = row.get(f.getName());
+                f.set(mappedObject, value);
+            }
+        }
+        return mappedObject;
+    }
+
+    private HashMap getMap(ResultSet rs) throws SQLException, ClassNotFoundException {
+        ResultSetMetaData meta_data = rs.getMetaData();
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:39]: rename to metaData

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/GenericQueryBuilder.java [hunk 8]
```diff
@@ -65,7 +60,7 @@ static List<String> getFieldsWithCondition(Object object, QueryFieldChecker chec
             field.setAccessible(true);
             try {
                 if(!field.getType().isPrimitive() && checkCondition.check(field, object)){
-                    if(isWrapper(field)){
+                    if(DataUtils.isWrapper(field)){
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/GenericQueryBuilder.java:63]: Rename to ObjectUtls

File: digit-health-services/libraries/health-services-common/src/test/java/org/digit/health/common/data/mapper/GenericRowMapperTest.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    }
+}
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
+@Builder
+class Employee{
+    int id;
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/test/java/org/digit/health/common/data/mapper/GenericRowMapperTest.java:59]: Make all these instance variables as private

File: digit-health-services/libraries/health-services-common/src/test/java/org/digit/health/common/data/mapper/GenericRowMapperTest.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+@Getter
+@Setter
+@Builder
+class NestedEmployee{
+    int id;
+    String name;
+    Amount amount;
+}
+
+@AllArgsConstructor
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/test/java/org/digit/health/common/data/mapper/GenericRowMapperTest.java:74]: Test using different annotations

**Review by jayantp-egov — CHANGES_REQUESTED** [review]

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 1 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: Updated.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 2 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: Updated

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 3 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: Updated.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 4 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: Updated.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 5 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: Updated.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 6 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: Updated.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 7 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: updated.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/GenericQueryBuilder.java (hunk 8 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/GenericQueryBuilder.java]: Updated.

File: digit-health-services/libraries/health-services-common/src/test/java/org/digit/health/common/data/mapper/GenericRowMapperTest.java (hunk 10 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/test/java/org/digit/health/common/data/mapper/GenericRowMapperTest.java]: Updated.

**Comment by shivab-egov** [comment]
GenericRowMapper requires Default Constructor or Constructor with no arguments.

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * GenericRowMapper is an implementation of RowMapper<T>. It is used for mapping ResultSet from sql query to List<T><p>
+ * GenericRowMapper throws SQLException if there is an error related to ResultSet, db or query.<p>
+ * GenericRowMapper throws RuntimeException whenever there is an error while converting resultset to object.<p>
+ *<p>
+ *<p>
+ * Requirements to use GenericRowMapper<p>
+ * * Column name and member variable name must be similar. for example columnName "price" will be mapped to member variable named "price"<p>
+ * * Column type and member variable type must be compatible. for example columnType "int" should be mapped to numeric data type, if you try map "int" to string, etc. IllegalArgumentException will be thrown.<p>
+ * * DefaultConstructor needs to be present for the object. DefaultConstructor is a constructor with no arguments.<p>
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:24]: Put space between DefaultConstructor
Default Constructor

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.Map;
+
+/**
+ * GenericRowMapper is an implementation of RowMapper<T>. It is used for mapping ResultSet from sql query to List<T><p>
+ * GenericRowMapper throws SQLException if there is an error related to ResultSet, db or query.<p>
+ * GenericRowMapper throws RuntimeException whenever there is an error while converting resultset to object.<p>
+ *<p>
+ *<p>
+ * Requirements to use GenericRowMapper<p>
+ * * Column name and member variable name must be similar. for example columnName "price" will be mapped to member variable named "price"<p>
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java:22]: Make F capital for "for example" and put a comma after it.
"For example, "

**Review by jayantp-egov — CHANGES_REQUESTED** [review]

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/SelectQueryBuilder.java [hunk 13]
```diff
@@ -1,15 +1,25 @@
...
+
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 public class SelectQueryBuilder implements GenericQueryBuilder {
 
+    Map<String, Object> paramsMap = new HashMap<>();
+
+    public Map<String, Object> getMap(){
```
Comment (jayantp-egov) [inline digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/SelectQueryBuilder.java:13]: Rename to getParamsMap

**Review by jayantp-egov — CHANGES_REQUESTED** [review]

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/SelectQueryBuilder.java (hunk 13 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/query/builder/SelectQueryBuilder.java]: Updated

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 12 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: Updated

File: digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java (hunk 11 above)
Comment (shivab-egov) [reply digit-health-services/libraries/health-services-common/src/main/java/org/digit/health/common/data/mapper/GenericRowMapper.java]: Updated.

**Review by jayantp-egov — APPROVED** [review]

**Pull request merged**